src/common/telemetry/lib/Telemetry.cpp

```diff
@@ -0,0 +1,365 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "Telemetry.hpp"
+#include "ParameterSets.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <iostream>
+#include <thread>
+
+using namespace MAT;
+
+LOGMANAGER_INSTANCE
+
+namespace Telemetry
+{
+
+OsConfigLogHandle TelemetryManager::m_log = NULL;
+MAT::ILogger* TelemetryManager::m_logger = nullptr;
+bool TelemetryManager::m_initialized = false;
+
+TelemetryManager::~TelemetryManager() noexcept
+{
+    Shutdown();
+}
+
+void TelemetryManager::SetupConfiguration(bool enableDebug, int teardownTime)
+{
+    auto& logConfig = LogManager::GetLogConfiguration();
+
+    logConfig[CFG_BOOL_ENABLE_TRACE] = enableDebug;
+    logConfig[CFG_INT_TRACE_LEVEL_MIN] = 0;
+    logConfig[CFG_INT_MAX_TEARDOWN_TIME] = teardownTime;
+
+    m_logger = LogManager::Initialize(API_KEY);
+}
+
+bool TelemetryManager::Initialize(bool enableDebug, int teardownTime)
+{
+    if (m_initialized)
+    {
+        return true;
+    }
+
+    OsConfigLogInfo(m_log, "Initializing telemetry...");
+
+    try
+    {
+        if (NULL == m_log)
+        {
+            m_log = OpenLog("/var/log/osconfig_telemetry.log", NULL);
+        }
+
+        SetupConfiguration(enableDebug, teardownTime);
+
+        if (!m_logger)
+        {
+            return false;
+        }
+
+        LogManager::SetTransmitProfile(TransmitProfile::TransmitProfile_RealTime);
+        m_initialized = true;
+        OsConfigLogInfo(m_log, "Telemetry initialized successfully.");
+        return true;
+    }
+    catch (const std::exception& e)
+    {
+        if (m_log)
```

Comment:
this seems wrong to me as you can throw and leave the members vars half set. Avoid this pattern either you construct and initialize of not.